data-access-layer/eventmesh/ApiServerClient.js

```diff
@@ -488,6 +488,75 @@ class ApiServerClient {
       });
   }
 
+  createConfigMapResource(configName, configParam) {
+    logger.info(`Creating ConfigMap ${configName} with data: ${configParam}`);
+    const metadata = {
+      name: configName
+    };
+    let data = {};
+    data = _.set(data, configParam.key, configParam.value);
+    const resourceBody = {
+      apiVersion: CONST.APISERVER.CONFIG_MAP.API_VERSION,
+      kind: CONST.APISERVER.CONFIG_MAP.RESOURCE_KIND,
+      metadata: metadata,
+      data: data
+    };
+    return Promise.try(() => this.init())
+      .then(() => apiserver.api[CONST.APISERVER.CONFIG_MAP.API_VERSION]
+        .namespaces(CONST.APISERVER.NAMESPACE)[CONST.APISERVER.CONFIG_MAP.RESOURCE_TYPE].post({
+          body: resourceBody
+        }))
+      .catch(err => {
+        return convertToHttpErrorAndThrow(err);
+      });
+  }
+
+  getConfigMapResource(configName) {
+    logger.debug('Get resource with opts: ', configName);
+    return Promise.try(() => this.init())
+      .then(() => apiserver.api[CONST.APISERVER.CONFIG_MAP.API_VERSION]
+        .namespaces(CONST.APISERVER.NAMESPACE)[CONST.APISERVER.CONFIG_MAP.RESOURCE_TYPE](configName).get())
+      .then(resource => {
+        return resource.body;
+      })
+      .catch(err => {
+        return convertToHttpErrorAndThrow(err);
+      });
+  }
+
+  createUpdateConfigMapResource(configName, configParam) {
+    const metadata = {
+      name: configName
+    };
+    let data = {};
+    data = _.set(data, configParam.key, configParam.value);
+    const resourceBody = {
+      apiVersion: CONST.APISERVER.CONFIG_MAP.API_VERSION,
+      kind: CONST.APISERVER.CONFIG_MAP.RESOURCE_KIND,
+      metadata: metadata,
+      data: data
+    };
+    return Promise.try(() => this.init())
+      .then(() => this.getConfigMapResource(configName))
+      .then((oldResourceBody) => {
+        resourceBody.data = oldResourceBody.data ? _.merge(oldResourceBody.data, data) : resourceBody.data;
+        return apiserver.api[CONST.APISERVER.CONFIG_MAP.API_VERSION]
```

Comment:
In my view, this can lead to race conditions. Try patching with `resourceVersion` also in resource body.

Race condition:
=> Thread 1 => Get config map (c1)
=> Thread 2 => Update it with some random data ()
=> Thread 1 => Update config map c1
Now without resource Version in c1 `body.metadata` 3rd call will overwrite whatever changes done by 2nd call.
If you pass resource version of c1 in 3rd call, it will through `409` if object has been changed meanwhile (Here by 2nd call)